File: plugins/plugin-im/sys_message/conversation.go
package sys_message

import (
	"context"
	"sort"
	"time"

	"hei-gin/sdk/db"
	"hei-gin/sdk/enums"
	"hei-gin/sdk/exception"
	"hei-gin/sdk/pojo"
	imModel "hei-gin/plugins/plugin-im/model"

	sysUser "hei-gin/plugins/plugin-sys/user"
	cliUser "hei-gin/plugins/plugin-client/user"
	"hei-gin/plugins/plugin-im/group"
)

// ==================== Conversations ====================

func Conversations(currentUserID, userType string, cursor string, size int) ([]ConversationVO, bool) {
	if size < 1 {
		size = 20
	}
	if size > 100 {
		size = 100
	}

	// Get conversations from im_conversation table
	var convs []imModel.Conversation
	q := db.DB.Model(&imModel.Conversation{}).
		Where("(user_id_1 = ? AND user_type_1 = ?) OR (user_id_2 = ? AND user_type_2 = ?)",
			currentUserID, userType, currentUserID, userType)
	if cursor != "" {
		if t, err := time.Parse("2006-01-02 15:04:05", cursor); err == nil {
			q = q.Where("last_time < ?", t)
		}
	}
	q.Order("last_time DESC").Limit(size + 1).Find(&convs)

	hasMore := len(convs) > size
	if hasMore {
		convs = convs[:size]
	}

	// Build single-chat conversation VOs
	type otherUser struct {
		ID   string
		Type string
	}
	others := make([]otherUser, 0, len(convs))
	resultMap := make(map[string]*ConversationVO, len(convs))

	for _, conv := range convs {
		var otherID, otherType string
		if conv.UserID1 == currentUserID && conv.UserType1 == userType {
			otherID, otherType = conv.UserID2, conv.UserType2
		} else {
			otherID, otherType = conv.UserID1, conv.UserType1
		}
		others = append(others, otherUser{ID: otherID, Type: otherType})
		lastTime := ""
		if conv.LastTime != nil {
			lastTime = pojo.FormatDateTime(*conv.LastTime)
		}
		resultMap[conv.ID] = &ConversationVO{
			ConversationID:   conv.ID,
			ConversationType: ConvTypeSingle,
			OtherUserID:      otherID,
			OtherUserType:    otherType,
			LastContent:      conv.LastMsg,
			LastTime:         lastTime,
		}
	}

	// Batch resolve nicknames/avatars
	businessIDs, consumerIDs := []string{}, []string{}
	for _, o := range others {
		switch o.Type {
		case string(enums.LoginTypeBusiness):
			businessIDs = append(businessIDs, o.ID)
		case string(enums.LoginTypeConsumer):
			consumerIDs = append(consumerIDs, o.ID)
		}
	}
	nicknameMap := make(map[string]string)
	avatarMap := make(map[string]string)
	if len(businessIDs) > 0 {
		var busUsers []sysUser.SysUser
		db.DB.Model(&sysUser.SysUser{}).Where("id IN ?", businessIDs).Find(&busUsers)
		for _, u := range busUsers {
			if u.Nickname != nil {
				nicknameMap["BUSINESS:"+u.ID] = *u.Nickname
			}
			if u.Avatar != nil {
				avatarMap["BUSINESS:"+u.ID] = *u.Avatar
			}
		}
	}
	if len(consumerIDs) > 0 {
		var conUsers []cliUser.ClientUser
		db.DB.Model(&cliUser.ClientUser{}).Where("id IN ?", consumerIDs).Find(&conUsers)
		for _, u := range conUsers {
			if u.Nickname != nil {
				nicknameMap["CONSUMER:"+u.ID] = *u.Nickname
			}
			if u.Avatar != nil {
				avatarMap["CONSUMER:"+u.ID] = *u.Avatar
			}
		}
	}
	for _, vo := range resultMap {
		key := vo.OtherUserType + ":" + vo.OtherUserID
		vo.OtherNickname = nicknameMap[key]
		vo.OtherAvatar = avatarMap[key]
	}

	// Get unread counts
	for _, vo := range resultMap {
		var cnt int64
		db.DB.Model(&imModel.Message{}).
			Where("conversation_id = ? AND receiver_id = ? AND status = ?", vo.ConversationID, currentUserID, "unread").
			Count(&cnt)
		vo.UnreadCount = cnt
	}

	// Add group conversations
	groupConvs := group.MyGroupConversations(currentUserID, userType)
	for _, gv := range groupConvs {
		resultMap["group:"+gv.GroupID] = &ConversationVO{
			ConversationID:   "group:" + gv.GroupID,
			ConversationType: ConvTypeGroup,
			GroupID:          gv.GroupID,
			GroupName:        gv.GroupName,
			GroupAvatar:      gv.GroupAvatar,
			MemberCount:      gv.MemberCount,
			LastContent:      gv.LastContent,
			LastTime:         gv.LastTime,
			UnreadCount:      gv.UnreadCount,
		}
	}

	// Convert to sorted slice
	result := make([]ConversationVO, 0, len(resultMap))
	for _, vo := range resultMap {
		result = append(result, *vo)
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].LastTime > result[j].LastTime
	})
	return result, hasMore
}

// ==================== Conversation Messages ====================

func BusinessConversationMessages(ctx context.Context, currentUserID string, conversationID string, cursor string, size int) ([]ConversationMessageVO, bool) {
	return conversationMessages(ctx, currentUserID, conversationID, cursor, size)
}

func ConsumerConversationMessages(ctx context.Context, currentUserID string, conversationID string, cursor string, size int) ([]ConversationMessageVO, bool) {
	return conversationMessages(ctx, currentUserID, conversationID, cursor, size)
}

func conversationMessages(ctx context.Context, currentUserID string, conversationID string, cursor string, size int) ([]ConversationMessageVO, bool) {
	if size < 1 {
		size = 20
	}
	if size > 100 {
		size = 100
	}

	q := db.DB.WithContext(ctx).Model(&imModel.Message{}).
		Where("conversation_id = ? AND (sender_id = ? OR receiver_id = ?) AND (deleted_by != ? OR deleted_by IS NULL)",
			conversationID, currentUserID, currentUserID, currentUserID)
	if cursor != "" {
		if t, err := time.Parse("2006-01-02 15:04:05", cursor); err == nil {
			q = q.Where("created_at < ?", t)
		}
	}
	var records []imModel.Message
	q.Order("created_at DESC").Limit(size + 1).Find(&records)

	hasMore := len(records) > size
	if hasMore {
		records = records[:size]
	}

	result := make([]ConversationMessageVO, len(records))
	for i, m := range records {
		result[i] = ConversationMessageVO{
			ID: m.ID, SenderID: m.SenderID, SenderType: m.SenderType,
			Content: m.Content, MsgType: m.MsgType, Extra: m.Extra,
			Status: m.Status,
			CreatedAt: pojo.FormatDateTimePtr(m.CreatedAt),
		}
	}
	return result, hasMore
}

// ==================== Get or Create Conversation ====================

func GetOrCreateConversation(userID string, userType string, param *GetOrCreateConversationParam) (string, string) {
	if param.UserID == "" || param.UserType == "" {
		panic(exception.NewBusinessError("参数错误", 400))
	}
	cid := imModel.GenerateConversationID(userID, enums.LoginTypeEnum(userType), param.UserID, enums.LoginTypeEnum(param.UserType))
	// Ensure the conversation cache exists
	var conv imModel.Conversation
	err := db.DB.Where("id = ?", cid).First(&conv).Error
	if err != nil {
		now := time.Now()
		conv = imModel.Conversation{
			ID:        cid,
			UserID1:   userID,
			UserType1: userType,
			UserID2:   param.UserID,
			UserType2: param.UserType,
			CreatedAt: &now,
		}
		db.DB.Create(&conv)
	}

	// Resolve display name
	displayName := param.UserID
	if param.UserType == string(enums.LoginTypeBusiness) {
		var u sysUser.SysUser
		if err := db.DB.First(&u, "id = ?", param.UserID).Error; err == nil && u.Nickname != nil {
			displayName = *u.Nickname
		}
	} else {
		var u cliUser.ClientUser
		if err := db.DB.First(&u, "id = ?", param.UserID).Error; err == nil && u.Nickname != nil {
			displayName = *u.Nickname
		}
	}
	return cid, displayName
}

File: plugins/plugin_im/message/conversation.py
"""Conversation handling — single-chat + group unified list.

Mirrors hei-gin plugins/plugin-im/message/conversation.go.
"""

from __future__ import annotations

from datetime import datetime

from core.db import SessionLocal
from core.exception import BusinessException
from core.enums import LoginTypeEnum
from plugins.plugin_im.model.message import (
    generate_conversation_id,
)
from plugins.plugin_im.message.params import (
    ConversationVO, ConversationMessageVO, GetOrCreateConversationParam,
)
from plugins.plugin_im.group import my_group_conversations
from plugins.plugin_im.message.repository import MessageRepository
from plugins.plugin_im.message.user_repository import IMUserRepository

import logging
logger = logging.getLogger(__name__)


def _fmt_dt(dt) -> str:
    if dt is None:
        return ""
    return dt.strftime("%Y-%m-%d %H:%M:%S")


# ═════════════════════════════════════════════════════════════════════
# Conversations (unified: single-chat + group)
# ═════════════════════════════════════════════════════════════════════

def conversations(current_user_id: str, user_type: str, cursor: str = "", size: int = 20) -> tuple[list[ConversationVO], bool]:
    if size < 1:
        size = 20
    if size > 100:
        size = 100

    # Build single-chat conversations from im_message
    single_result = _build_from_messages(current_user_id, user_type)

    # Add group conversations
    group_convs = my_group_conversations(current_user_id, user_type)
    for gv in group_convs:
        key = "group:" + gv.group_id
        single_result[key] = ConversationVO(
            conversation_id=key,
            conversation_type="group",
            group_id=gv.group_id,
            group_name=gv.group_name,
            group_avatar=gv.group_avatar,
            member_count=gv.member_count,
            last_content=gv.last_content,
            last_time=gv.last_time,
            unread_count=gv.unread_count,
        )

    # Sort by last_time descending
    result = sorted(single_result.values(), key=lambda x: x.last_time, reverse=True)

    # Apply cursor pagination
    has_more = False
    if cursor:
        result = [vo for vo in result if vo.last_time > cursor]
    if len(result) > size:
        result = result[:size]
        has_more = True

    return result, has_more


def _build_from_messages(current_user_id: str, user_type: str) -> dict[str, ConversationVO]:
    db = SessionLocal()
    try:
        repository = MessageRepository(db)
        user_repository = IMUserRepository(db)
        rows = repository.list_latest_message_rows(current_user_id)
        if not rows:
            return {}

        conv_ids = [r.conversation_id for r in rows]
        unread_map = repository.unread_counts_by_conversation(conv_ids, current_user_id)
        business_ids = []
        consumer_ids = []
        result_map = {}

        for r in rows:
            if r.sender_id == current_user_id and r.sender_type == user_type:
                other_id, other_type = r.receiver_id, r.receiver_type
            else:
                other_id, other_type = r.sender_id, r.sender_type

            if other_type == "BUSINESS":
                business_ids.append(other_id)
            else:
                consumer_ids.append(other_id)

            result_map[r.conversation_id] = ConversationVO(
                conversation_id=r.conversation_id,
                conversation_type="single",
                other_user_id=other_id,
                other_user_type=other_type,
                last_content=r.content or "",
                last_time=_fmt_dt(r.created_at),
                unread_count=unread_map.get(r.conversation_id, 0),
            )

        nickname_map = {}
        avatar_map = {}
        if business_ids:
            users = user_repository.list_sys_users(business_ids)
            for u in users:
                nickname_map["BUSINESS:" + u.id] = u.nickname or u.username or ""
                avatar_map["BUSINESS:" + u.id] = u.avatar or ""
        if consumer_ids:
            users = user_repository.list_client_users(consumer_ids)
            for u in users:
                nickname_map["CONSUMER:" + u.id] = u.nickname or u.username or ""
                avatar_map["CONSUMER:" + u.id] = u.avatar or ""
        for vo in result_map.values():
            key = vo.other_user_type + ":" + vo.other_user_id
            vo.other_nickname = nickname_map.get(key, "")
            vo.other_avatar = avatar_map.get(key, "")

        return result_map
    finally:
        db.close()


# ═════════════════════════════════════════════════════════════════════
# Conversation Messages
# ═════════════════════════════════════════════════════════════════════

def conversation_messages(
    current_user_id: str, conversation_id: str, cursor: str = "", size: int = 20
) -> tuple[list[ConversationMessageVO], bool]:
    if size < 1:
        size = 20
    if size > 100:
        size = 100

    db = SessionLocal()
    try:
        repository = MessageRepository(db)
        cursor_dt = None
        if cursor:
            try:
                cursor_dt = datetime.strptime(cursor, "%Y-%m-%d %H:%M:%S")
            except ValueError:
                pass
        records = repository.list_conversation_messages(current_user_id, conversation_id, cursor_dt, size)
        has_more = len(records) > size
        if has_more:
            records = records[:size]

        result = []
        for m in records:
            file_url = ""
            if m.msg_type in ("IMAGE", "FILE"):
                from plugins.plugin_im.message.im_file import resolve_file_url
                file_url = resolve_file_url(m.content or "", m.extra or "")
            result.append(ConversationMessageVO(
                id=m.id,
                sender_id=m.sender_id or "",
                sender_type=m.sender_type or "",
                content=m.content or "",
                msg_type=m.msg_type,
                extra=m.extra or "",
                status=m.status,
                file_url=file_url,
                created_at=_fmt_dt(m.created_at),
            ))
        return result, has_more
    finally:
        db.close()


# ═════════════════════════════════════════════════════════════════════
# Get or Create Conversation
# ═════════════════════════════════════════════════════════════════════

def get_or_create_conversation(user_id: str, user_type: str, param: GetOrCreateConversationParam) -> tuple[str, str]:
    if not param.user_id or not param.user_type:
        raise BusinessException("参数错误", 400)

    cid = generate_conversation_id(
        user_id, LoginTypeEnum(user_type),
        param.user_id, LoginTypeEnum(param.user_type),
    )

    display_name = param.user_id
    db = SessionLocal()
    try:
        user_repository = IMUserRepository(db)
        if param.user_type == "BUSINESS":
            u = user_repository.find_sys_user(param.user_id)
            if u:
                display_name = u.nickname or u.username or param.user_id
        else:
            u = user_repository.find_client_user(param.user_id)
            if u:
                display_name = u.nickname or u.username or param.user_id
    finally:
        db.close()

    return cid, display_name
